devdoc/interlocked_linux_requirements.md

```diff
@@ -0,0 +1,345 @@
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile int32_t*, destination, int32_t, exchange, int32_t, comperand);
+MOCKABLE_FUNCTION(, int16_t, interlocked_compare_exchange_16, volatile int16_t*, destination, int16_t, exchange, int16_t, comperand);
+MOCKABLE_FUNCTION(, int64_t, interlocked_compare_exchange_64, volatile int64_t*, destination, int64_t, exchange, int64_t, comperand);
+MOCKABLE_FUNCTION(, void*, interlocked_compare_exchange_pointer, void* volatile*, destination, void*, exchange, void*, comperand);
+MOCKABLE_FUNCTION(, int32_t, interlocked_decrement, volatile int32_t*, addend);
+MOCKABLE_FUNCTION(, int16_t, interlocked_decrement_16, volatile int16_t*, addend);
+MOCKABLE_FUNCTION(, int64_t, interlocked_decrement_64, volatile int64_t*, addend);
+MOCKABLE_FUNCTION(, int32_t, interlocked_exchange, volatile int32_t*, target, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_exchange_16, volatile int16_t*, target, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_exchange_64, volatile int64_t*, target, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_exchange_8, volatile int8_t*, target, int8_t, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_exchange_add, volatile int32_t*, addend, int32_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_exchange_add_64, volatile int64_t*, addend, int64_t, value);
+MOCKABLE_FUNCTION(, void*, interlocked_exchange_pointer, void* volatile*, target, void*, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_increment, volatile int32_t*, addend);
+MOCKABLE_FUNCTION(, int16_t, interlocked_increment_16, volatile int16_t*, addend);
+MOCKABLE_FUNCTION(, int64_t, interlocked_increment_64, volatile int64_t*, addend);
+MOCKABLE_FUNCTION(, int32_t, interlocked_or, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_or_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_or_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_or_8, volatile int8_t*, destination, int8_t, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_xor, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_xor_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_xor_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_xor_8, volatile int8_t*, destination, int8_t, value);
+```
+
+## interlocked_add
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_001: [** `interlocked_add` shall call `atomic_fetch_add` from `stdatomic.h`. **]**

```

Comment:
>atomic_fetch_add [](start = 66, length = 16)

So for InterlockedX calls, you didn't need this because you were passing all arguments "as is". Since the atomic_X functions have slightly different parameters, we should follow the convention we use in other specs. So this requirement should be like:

`interlocked_add` shall call `atomic_fetch_add` with `addend` as `object` and `value` as `operand`.

Things like interlocked compare exchange use different orders of arguments, so it's nice to be very explicit here.

Reply:
Here and everywhere in this file

---
In reply to: [431449568](https://github.com/Azure/azure-c-util/pull/44#discussion_r431449568) [](ancestors = 431449568)